src/cart/src/gurt/hash.c

```diff
@@ -352,298 +398,351 @@ ch_rec_insert(struct d_hash_table *htable, unsigned idx, d_list_t *rlink)
 #endif
 }
 
+/**
+ * Insert the record into the hash table and take refcount on it if
+ * "ephemeral" is not set.
+ */
+static inline void
+ch_rec_insert_addref(struct d_hash_table *htable, struct d_hash_bucket *bucket,
+		     d_list_t *link)
+{
+	if (!(htable->ht_feats & D_HASH_FT_EPHEMERAL))
+		ch_rec_addref(htable, link);
+
+	ch_rec_insert(htable, bucket, link);
+}
+
 static inline void
-ch_rec_delete(struct d_hash_table *htable, d_list_t *rlink)
+ch_rec_delete(struct d_hash_table *htable, d_list_t *link)
 {
-	d_list_del_init(rlink);
+	d_list_del_init(link);
```

Comment:
does this need to use del_init vs del ?  The only reason del_init is ever needed is if you turn around and do d_list_empty on the link or some such.  Otherwise, if added to a new list, it will be overwritten anyway.

Reply:
we have a **d_hash_rec_unlinked()** which exactly check this.